### PR TITLE
Naji/proxy peering integration tests

### PR DESCRIPTION
**Pull request by NajiObeid**
https://github.com/gravitational/cloud/issues/1475

### CODE REVIEW

File: integration/proxy_tunnel_strategy_test.go
Comment (dboslee) [inline integration/proxy_tunnel_strategy_test.go:144]: We discussed this the other day but figured we could discuss it more here. 

Does the peer to peer grpc connection cause the test to be flaky? Meaning if the client connection and dial request go through before the grpc connection is setup could this test fail? 

From what I remember we saw the test passing reliably. I think we should just double check that it is indeed reliable. I think the thing to validate is what happens if we use a grpc clientConn that is not established yet. Either it fails immediately in which case this could be flaky or it waits with some timeout for the connection to be established in which case we shouldn't need to do anything else.

Comment (dboslee) [inline integration/proxy_tunnel_strategy_test.go:144]: We discussed this the other day but figured we could discuss it more here. 

Does the peer to peer grpc connection cause the test to be flaky? Meaning if the client connection and dial request go through before the grpc connection is setup could this test fail? 

From what I remember we saw the test passing reliably. I think we should just double check that it is indeed reliable. I think the thing to validate is what happens if we use a grpc clientConn that is not established yet. Either it fails immediately in which case this could be flaky or it waits with some timeout for the connection to be established in which case we shouldn't need to do anything else.

File: integration/proxy_tunnel_strategy_test.go
Comment (rosstimothy) [inline integration/proxy_tunnel_strategy_test.go:213]:
```suggestion
	lb, err := utils.NewLoadBalancer(context.Background(), *lbAddr)
```

File: integration/proxy_tunnel_strategy_test.go
Comment (rosstimothy) [inline integration/proxy_tunnel_strategy_test.go:407]:
```suggestion
		require.NoError(t, postgresDB.Close())
```

File: integration/proxy_tunnel_strategy_test.go
Comment (rosstimothy) [inline integration/proxy_tunnel_strategy_test.go:378]:
```suggestion
		require.NoError(t, db.StopAll())
```

File: integration/proxy_tunnel_strategy_test.go
Comment (rosstimothy) [inline integration/proxy_tunnel_strategy_test.go:326]:
```suggestion
		require.NoError(t, node.StopAll())
```

File: integration/proxy_tunnel_strategy_test.go
Comment (rosstimothy) [inline integration/proxy_tunnel_strategy_test.go:297]:
```suggestion
		require.NoError(t, proxy.StopAll())
```

File: integration/proxy_tunnel_strategy_test.go
Comment (rosstimothy) [inline integration/proxy_tunnel_strategy_test.go:256]:
```suggestion
		require.NoError(t, auth.StopAll())
```

File: integration/proxy_tunnel_strategy_test.go
Comment (rosstimothy) [inline integration/proxy_tunnel_strategy_test.go:217]:
```suggestion
		require.NoError(t, lb.Close())
```

File: integration/proxy_tunnel_strategy_test.go
Comment (NajiObeid) [reply integration/proxy_tunnel_strategy_test.go:148]: [36a4b7e](https://github.com/gravitational/teleport/pull/11845/commits/36a4b7ee23f35743bb9c7f88b492f6bc2c679741)

File: lib/proxy/client.go
Comment (dboslee) [inline lib/proxy/client.go:402]: Should this acquire a read lock?

File: integration/proxy_tunnel_strategy_test.go
Comment (jimbishopp) [inline integration/proxy_tunnel_strategy_test.go:59]:
```suggestion
	t.Cleanup(func() { lib.SetInsecureDevMode(false) })
```

File: integration/proxy_tunnel_strategy_test.go
Comment (jimbishopp) [inline integration/proxy_tunnel_strategy_test.go:103]:
```suggestion
	t.Cleanup(func() { lib.SetInsecureDevMode(false) })
```

File: integration/proxy_tunnel_strategy_test.go
Comment (jimbishopp) [inline integration/proxy_tunnel_strategy_test.go:231]: Can we check for `p.lb != nil` at the beginning of the function and change from `t.Error` to `require.Fail`?

File: integration/proxy_tunnel_strategy_test.go
Comment (jimbishopp) [inline integration/proxy_tunnel_strategy_test.go:268]: Same: check for `p.auth != nil` at the beginning of the function and change from `t.Error` to `require.Fail`.

File: integration/proxy_tunnel_strategy_test.go
Comment (jimbishopp) [inline integration/proxy_tunnel_strategy_test.go:338]: Same: check for `p.node != nil` at the beginning of the function and change from `t.Error` to `require.Fail`.

File: integration/proxy_tunnel_strategy_test.go
Comment (jimbishopp) [inline integration/proxy_tunnel_strategy_test.go:421]: Same: check for `p.db != nil` at the beginning of the function and change from `t.Error` to `require.Fail`.

**Review by jimbishopp — COMMENTED** [review]
A few minor comments.

File: integration/integration_test.go
Comment (NajiObeid) [reply integration/integration_test.go]: Yeah I didn't give it much of a thought and copied the function from above.
Hope the fix in [310dfd6](https://github.com/gravitational/teleport/pull/11845/commits/310dfd6ae368d20fdead24af5b372732d54a642b) won't confuse people more about why it even exists

Comment (NajiObeid) [reply integration/integration_test.go]: and I still haven't figure out a better name fort that function

**Review by dboslee — APPROVED** [review]

File: integration/helpers.go
Comment (NajiObeid) [reply integration/helpers.go:1895]: referencing a previous comment by @jimbishopp https://github.com/gravitational/teleport/pull/11845#discussion_r860327502

Error messages returned by `require.Eventually` and `require.Never` cannot be customized with variables that get updated during condition checks. So in order to keep using those functions we would have to lose some of the verbosity or alternatively come up with our own implementation.

Comment (zmb3) [reply integration/helpers.go:1895]: I think I would still prefer using `require.Eventually` rather than reimplementing it. Can always use `t.Log()` to log some extra context if the check fails.

File: integration/integration_test.go
Comment (NajiObeid) [reply integration/integration_test.go:3043]: I agree, but I honestly can't think of any other straightforward way of doing it. That also applies for the wait time, I couldn't figure out how much is too much or too little. I am open to suggestions.

File: integration/integration_test.go
Comment (NajiObeid) [reply integration/integration_test.go:3025]: Same answer as above, the numbers I have are mostly inspired by other numbers found in the same file, there's no reasoning behind them. I am open to any suggestions.

File: integration/integration_test.go
Comment (NajiObeid) [reply integration/integration_test.go:3043]: The original intention for this was to make sure the env is setup the way it is intended, and not to test the proxy or reversetunnel lib directly

@dboslee is that something covered in tests for the agent changes you made?

I worry for possible regressions in the future if we managed to reach a node through a direct proxy connection and not through a proxy tunnel, tests would still pass and we wouldn't know.

**Comment by jimbishopp** [comment]
@NajiObeid @dboslee What work remains before this is good to merge?

**Comment by NajiObeid** [comment]
@jimbishopp tests are flaky at the moment and we're still investigating the root cause

**Comment by NajiObeid** [comment]
@rosstimothy @espadolini We got the flakiness sorted. Would love a second review. Thanks

**Review by jimbishopp — COMMENTED** [review]
First pass. I like the refactor. 👍

File: integration/integration_test.go
Comment (espadolini) [inline integration/integration_test.go:3037]: style nit: I preferred the previous style of indentation

File: lib/proxy/server.go
Comment (espadolini) [inline lib/proxy/server.go:153]: Do we want to use `utils.IsUseOfClosedNetworkError(err)` here, for consistency with the rest of the codebase?

File: lib/reversetunnel/localsite.go
Comment (espadolini) [inline lib/reversetunnel/localsite.go:434]: What is this supposed to prevent?

File: integration/proxy_tunnel_strategy_test.go
Comment (espadolini) [inline integration/proxy_tunnel_strategy_test.go:293]: This call to `CreateEx` will redefine `conf.Proxy.WebAddr` (and the other listeners) and will add the `proxy` `TeleInstance`'s intended internal auth server's address to `conf.AuthServers`.

Also, AIUI nothing ever enables multiplex mode - not sure if we care about that here, but it's weird that we're running a single loadbalancer just for the web port while everything else is split between proxies.

File: integration/proxy_tunnel_strategy_test.go
Comment (espadolini) [inline integration/proxy_tunnel_strategy_test.go:323]: Same configuration caveat as the proxy one, `AuthServers` will contain the load balancer and the `node`'s internal (deactivated) auth server.

File: integration/proxy_tunnel_strategy_test.go
Comment (espadolini) [inline integration/proxy_tunnel_strategy_test.go:372]: Same configuration caveat as the proxy one, `AuthServers` will contain the load balancer and the `db`'s internal (deactivated) auth server.

File: integration/proxy_tunnel_strategy_test.go
Comment (espadolini) [inline integration/proxy_tunnel_strategy_test.go:420]: Can we access (the generic equivalent of) `.Metadata.Name` and `.Spec.ProxyIDs` through `types.Server` instead? Just so we have less changes to chase if we ever changed the concrete type to something else in the future.

**Review by espadolini — CHANGES_REQUESTED** [review]

File: integration/proxy_tunnel_strategy_test.go
Comment (dboslee) [inline integration/proxy_tunnel_strategy_test.go:441]:
```suggestion
			node, ok := nodeT.(types.Server)
			if !ok {
				return false, trace.BadParameter("invalid type: wanted Server got '%T'", nodeT)
			}
```

File: integration/proxy_tunnel_strategy_test.go
Comment (dboslee) [inline integration/proxy_tunnel_strategy_test.go:470]:
```suggestion
			db, ok := dbT.(types.DatabaseServer)
			if !ok {
				return false, trace.BadParameter("invalid type: wanted DatabaseServer got '%T'", dbT)
			}
```

**Review by dboslee — APPROVED** [review]

File: integration/proxy_tunnel_strategy_test.go
Comment (NajiObeid) [reply integration/proxy_tunnel_strategy_test.go:441]: now that you mention it, we don't need that type assertion at all. thanks

File: integration/proxy_tunnel_strategy_test.go
Comment (NajiObeid) [reply integration/proxy_tunnel_strategy_test.go:470]: same as above

**Review by espadolini — APPROVED** [review]

**Review by rosstimothy — APPROVED** [review]

**Pull request merged**
